setup/src/setup-haskell.ts

```diff
@@ -21,34 +29,106 @@ export default async function run(
   try {
     core.info('Preparing to setup a Haskell environment');
     const os = process.platform as OS;
+    const home = process.env.HOME;
     const opts = getOpts(getDefaults(os), os, inputs);
 
+    // Cache parameters
+    let cachePaths: string[] = [];
+    let hashFile = '';
+
     for (const [t, {resolved}] of Object.entries(opts).filter(o => o[1].enable))
       await core.group(`Installing ${t} version ${resolved}`, async () =>
         installTool(t as Tool, resolved, os)
       );
 
-    if (opts.stack.setup)
+    if (opts.stack.setup) {
       await core.group('Pre-installing GHC with stack', async () =>
         exec('stack', ['setup', opts.ghc.resolved])
       );
+    }
 
-    if (opts.cabal.enable)
+    if (opts.stack.enable) {
+      const stackRoot = os === 'win32' ? 'C:\\sr' : `${home}/.stack`;
+      core.setOutput('stack-root', stackRoot);
+      if (os === 'win32') core.exportVariable('STACK_ROOT', 'C:\\sr');
+
+      if (opts.cache) {
+        const matches = await glob
+          .create('**/stack.*.lock')
+          .then(async g => g.glob());
+        if (!matches[0]) await exec('stack', ['ls', 'dependencies']);
+        cachePaths = [stackRoot];
+        hashFile = matches[0] || 'stack.yaml.lock';
+      }
+    }
+
+    if (opts.cabal.enable) {
       await core.group('Setting up cabal', async () => {
         await exec('cabal', ['user-config', 'update'], {silent: true});
         const configFile = await cabalConfig();
 
-        if (process.platform === 'win32') {
+        if (os === 'win32')
           fs.appendFileSync(configFile, 'store-dir: C:\\sr\n');
-          core.setOutput('cabal-store', 'C:\\sr');
-        } else {
-          core.setOutput('cabal-store', `${process.env.HOME}/.cabal/store`);
-        }
 
         await exec('cabal user-config update');
         if (!opts.stack.enable) await exec('cabal update');
       });
+      const cabalStore = os === 'win32' ? 'C:\\sr' : `${home}/.cabal/store`;
+      core.setOutput('cabal-store', cabalStore);
+
+      if (opts.cache) {
+        const matches = await glob
+          .create('**/cabal.*.freeze')
+          .then(async g => g.glob());
+        if (!matches[0]) await exec('cabal freeze');
+        cachePaths = [cabalStore, 'dist-newstyle'];
+        hashFile = matches[0] || 'cabal.project.freeze';
+      }
+    }
+
+    if (opts.cache) {
+      core.info(`Loading cache...`);
+      const keys = [
+        os,
+        opts.ghc.resolved,
+        await hashFiles(hashFile),
+        process.env.GITHUB_SHA
+      ];
+      const st: CacheState = {
+        paths: opts.cachePaths || cachePaths,
+        keys: opts.cacheKeys || [
+          keys.join('-'),
+          keys.slice(0, 4).join('-') + '-',
+          keys.slice(0, 3).join('-') + '-',
+          keys.slice(0, 2).join('-') + '-'
+        ]
```

Comment:
You should be careful about the caching growing unbounded as time goes on. More info here: https://markkarpov.com/post/github-actions-for-haskell-ci.html#comment-6035135802

Reply:
The cache is designed to grow unbounded because github actions automatically limits it to the last 5G in size. It might theoretically make sense to let people override the key but if they want a very specific set of behavior, they'll probably end up writing their own usage of caching anyway.